.circleci/config.yml

```diff
@@ -0,0 +1,25 @@
+version: 2.1
+jobs:
+  build:
+    working_directory: ~/repo
+    docker:
+      - image: circleci/golang:1.15.8
+    steps:
+      - checkout
+      - restore_cache:
+          keys:
+            - go-mod-v4-{{ checksum "go.sum" }}
+      - run:
+          name: Install Dependencies
+          command: go mod download
+      - save_cache:
+          key: go-mod-v4-{{ checksum "go.sum" }}
+          paths:
+            - "/go/pkg/mod"
+      - run:
+          name: Build server and CLI
+          command: |
+            make build
+            make cli
```

Comment:
Why not do the 
```
make build
make cli
```
before the `save_cache` step?

Reply:
Because we don't want to save cache for those steps. We should be building the binaries on every CI run to make sure the code has no errors. It is supposed to pick up the new changes in the commits and build binaries for it.

Reply:
> save cache
the cache saving is happening only for the specific path `"/go/pkg/mod"`. Even if you're running `make build` or `make cli`, the additional packages downloaded while building would go in `"/go/pkg/mod"`. So the binaries built would still be new.

Reply:
Sorry I misunderstood the comment earlier, but why would be the benefit of building the binaries before cache step?

Reply:
```
      - restore_cache:
          keys:
            - go-mod-v4-{{ checksum "go.sum" }}
      - run:
          name: Install Dependencies
          command: go mod download
      - run:
          name: Build server and CLI
          command: |
            make build
            make cli
      - save_cache:
          key: go-mod-v4-{{ checksum "go.sum" }}
          paths:
            - "/go/pkg/mod"
```

We could do it in this way.

Reply:
Based on my understanding it's one and the same thing. We should be restoring the cache before `go mod download`, and saving the cache after it. It doesn't matter if it runs before or after the build step.